Add version catalog support for Gradle

alcere opened the pull request:
Hello from the GitHub [Android mobile team](https://github.com/mobile)!

As part of a client apps hackathon we decided to give it a shot and add [version catalog](https://docs.gradle.org/current/userguide/platforms.html) support for Gradle projects.

This functionality has been requested by some users here: https://github.com/dependabot/dependabot-core/issues/3121, and we are planning to start using version catalog as well as soon as Dependabot supports it :P.

We have added support for version catalogs defined using a `toml` file ([more info](https://docs.gradle.org/current/userguide/platforms.html#sub:conventional-dependencies-toml)), for libraries and plugins using value versions and referenced versions, so this changes add support for **a** toml file like:

```
[versions]
corektx = "1.7.0"
lifecycleRuntime = "2.3.1"
composeUi = "1.2.0"
jUnitVersion = "4.13.2"
androidxJunit = "1.1.3"
espresso = "3.5.0"

[libraries]
corektx = { module = "androidx.core:core-ktx", version.ref = "corektx" }
lifecycleRuntimeKtx = { module = "androidx.lifecycle:lifecycle-runtime-ktx", version.ref = "lifecycleRuntime" }

compose-activity = { module = "androidx.activity:activity-compose", version = "1.3.1" }
compose-ui = { module = "androidx.compose.ui:ui", version.ref = "composeUi" }
compose-ui-tooling = { module = "androidx.compose.ui:ui-tooling-preview", version.ref = "composeUi" }
compose-material = { module = "androidx.compose.material:material", version = { strictly = "[1.2.0, 1.3.1[", prefer="1.3.0" } }

test-junit = { module = "junit:junit", version.ref = "jUnitVersion" }
test-andoridx-junit = { module = "androidx.test.ext:junit", version.ref = "androidxJunit" }
test-compose-ui = { module = "androidx.compose.ui:ui-test-junit4", version.ref = "composeUi" }
test-compose-ui-tooling = { module = "androidx.compose.ui:ui-tooling", version.ref = "composeUi" }
test-compose-manifest = { module = "androidx.compose.ui:ui-test-manifest", version.ref = "composeUi" }
test-espresso-Core = { module = "androidx.test.espresso:espresso-core", version.ref = "espresso" }

[bundles]
compose = ["compose-activity", "compose-ui", "compose-ui-tooling", "compose-material"]

[plugins]
kotlinter = { id = "org.jmailen.kotlinter", version = "3.11.0" }
```

### Testing

 In order to guide our development we have used the steps defined in [here](https://github.com/dependabot/dependabot-core#dry-run-script) to use the dry-run script pointing to [this](https://github.com/bigandroidenergies/version_catalog) test repo we created.

### Note

We realize that this code won't reach the quality bar to be considered as a candidate to be merged (Ruby is not our field of expertise), but we will be more than happy to react to your feedback and keep working on it!

### BIG THANKS
To @jurre who has been helping us on this task, sharing advices, techniques and ideas!

Closes #3121.

trevjonez commented on gradle/lib/dependabot/gradle/file_fetcher.rb:
```diff
@@ -14,6 +14,9 @@ class FileFetcher < Dependabot::FileFetchers::Base
       SUPPORTED_SETTINGS_FILE_NAMES =
         %w(settings.gradle settings.gradle.kts).freeze
 
+      # For now Gradle only supports libray .toml files in the main gradle folder
+      SUPPORTED_VERSION_CATALOG_FILE_PATH = %w(/gradle/libs.versions.toml).freeze
```
That isn't always true. Using [`from` in the settings dsl](https://docs.gradle.org/current/javadoc/org/gradle/api/initialization/dsl/VersionCatalogBuilder.html#from-java.lang.Object-) you can import a toml catalog from any arbitrary path. `gradle/libs.versions.toml` is just the default convention that gradle will pickup without having to explicitly list it in your `settings.gradle[.kts]`

https://docs.gradle.org/current/userguide/platforms.html#sec:importing-catalog-from-file

I would suspect handling any toml file in the given [config directory](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#directory) would be safe? Or rather/maybe `root_dir/gradle/*.toml` + `root_dir/*.toml`? or maybe grep out lines using `from("*.toml")` to build the list of catalogs?

FWIW gradle does enforce that `from` can only be called once per catalog.
```kotlin
dependencyResolutionManagement {
    versionCatalogs {
        create("libs") {
            from("boom.versions.toml")
        }
    }
}
```
produces:
```text
Invalid catalog definition:
  - Problem: In version catalog libs, you can only call the 'from' method a single time.

* Try:
> Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.InvalidUserDataException: Invalid catalog definition:
  - Problem: In version catalog libs, you can only call the 'from' method a single time.
    
    Reason: The method was called more than once.
    
    Possible solution: Remove further usages of the method call.
    
    Please refer to https://docs.gradle.org/7.6/userguide/version_catalog_problems.html#too_many_import_invocation for more details about this problem.
	at org.gradle.api.internal.catalog.problems.DefaultCatalogProblemBuilder.maybeThrowError(DefaultCatalogProblemBuilder.java:66)
[...]
```

alcere replied:
Sorry about the confusion created by the comment here! Your explanation and concern are valid, and your proposal looks good too.
The initial goal for this PR was to keep things simple and only support the default location of the `.toml` file, not considering any `dependencyResolutionManagement` extra configuration, leaving that task as a follow-up piece of work.

lbenedetto replied:
So it wouldn't support declaring the version directly in the `settings.gradle` file then?

eg:
```
dependencyResolutionManagement {
    versionCatalogs {
        create("libs") {
            version("lombok", "1.18.24")
            library("lombok", "org.projectlombok", "lombok").versionRef("lombok")
        }
    }
}
```

deivid-rodriguez replied:
That's correct.

deivid-rodriguez replied:
Reticketed as https://github.com/dependabot/dependabot-core/issues/6831.

lbenedetto replied:
Also reticketed my request as #6838

trevjonez reviewed (commented):
As is this would cover the bulk of projects using version catalogs. But surely as things scale up and out we would see projects that have multiple catalogs or non conventional file layouts in use.

deivid-rodriguez commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -62,6 +67,79 @@ def self.find_includes(buildfile, dependency_files)
+          [nil, nil]
+        end
+        [group, name]
+      end
```
If I'm understanding the logic of this method correctly, you can write it a bit simpler

```ruby
def library_parse_declaration(declaration)
  if declaration["module"]
    declaration["module"].split(":")
  else
    [declaration["group"], declaration["name"]]
  end
end
```

alcere replied:
Totally, done here: https://github.com/dependabot/dependabot-core/pull/6249/commits/50f51153c301d296da0895b087a18090fb75d0e8

deivid-rodriguez commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -62,6 +67,79 @@ def self.find_includes(buildfile, dependency_files)
+
+          # Only support basic version and reference formats for now,
+          # refrain from updating anything else as it's likely to be a very deliberate choice.
+          next if (version.is_a?(Hash) && (version.length != 1 || version["ref"].nil?)) || version.include?("[")
```
Same comment as with dependencies, I guess making the same comment suggests that we could DRY up things a bit?

alcere replied:
Great idea! Done here: https://github.com/dependabot/dependabot-core/pull/6249/commits/e07e59b81d42155b5577956de0ed568c5c4e1f3e

deivid-rodriguez commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -62,6 +67,79 @@ def self.find_includes(buildfile, dependency_files)
+      def parsed_toml_file(file)
+        @parsed_file ||= {}
+        @parsed_file[file.name] ||= TomlRB.parse(file.content)
+      rescue TomlRB::ParseError, TomlRB::ValueOverwriteError
```
I would think `TomlRB::ValueOverwriteError` this would be an error when writing a Toml file, not when parsing it, but we're rescuing it everywhere together with the `TomlRB::ParseError`, so 👍.

deivid-rodriguez reviewed (commented):
This is pretty awesome! I made a first pass on reviewing.

My main concern is that when sharing version declarations between multiple dependencies, I think Dependabot may create a different PR for each dependency that uses the reference, all of them with the same contents? Is this correct?

bmarty commented on gradle/lib/dependabot/gradle/file_fetcher.rb:
```diff
@@ -82,6 +85,15 @@ def subproject_buildfiles(root_dir)
+
+        gradle_toml_file(root_dir)
+      rescue Dependabot::DependencyFileNotFound
+        # Catalog file it's optional for Gradle
```
```suggestion
        # Catalog file is optional for Gradle
```

AmsterdamFilho commented:
Really looking forward to this 🙏

dsvensson commented:
Would it be more representable to look at [the gradle lockfile](https://docs.gradle.org/current/userguide/dependency_locking.html)?

If I depend on `X` which has an outdated dependency `Y`, and I add an `runtimeOnly(libs.Y)` where `runtimeOnly(libs.X)` is used this PR will still think that the project still contains vulnerabilities?

This PR is however still a huge upgrade to the current state of things where I have to remember to generate a POM-file that's otherwise unused.

dsvensson commented:
Would it be more representable to look at [the gradle lockfile](https://docs.gradle.org/current/userguide/dependency_locking.html)? The `libs.versions.toml` file doesn't know about transitive dependencies (but maybe dependabot crawls that graph?), nor if these are updated [via gradle version overrides](https://docs.gradle.org/current/userguide/dependency_constraints.html).

trevjonez commented:
> Would it be more representable to look at [the gradle lockfile](https://docs.gradle.org/current/userguide/dependency_locking.html)? The `libs.versions.toml` file doesn't know about transitive dependencies (but maybe dependabot crawls that graph?), nor if these are updated [via gradle version overrides](https://docs.gradle.org/current/userguide/dependency_constraints.html).

FWIW: In the 8 years I have been using Gradle, I have never worked on a project that used Gradle lock files. I suspect this is more related to how mobile builds and deployments differ from a backend system CI/CD pipeline. Typically we run tools that scan and monitor the fully resolved tree before releasing. IE snyk.io

dsvensson commented:
> FWIW: In the 8 years I have been using Gradle, I have never worked on a project that used Gradle lock files. I suspect this is more related to how mobile builds and deployments differ from a backend system CI/CD pipeline. Typically we run tools that scan and monitor the fully resolved tree before releasing. IE snyk.io

This is because lockfiles was made available with 7.0 at 9 April 2021 so for 6 years and 2 months of those 8 years this feature wasn't available. Added to that, IDEA didn't really support it in the beginning, so maybe bump that to the first 7 of your 8 years. I know it's underused, but that doesn't change anything. It is the only way to avoid false positives, just wanted note that, could be a future improvement. Hopefully some future supply chain poisoning makes developers adopt lockfiles and dependency verification.

trevjonez commented:
> > FWIW: In the 8 years I have been using Gradle, I have never worked on a project that used Gradle lock files. I suspect this is more related to how mobile builds and deployments differ from a backend system CI/CD pipeline. Typically we run tools that scan and monitor the fully resolved tree before releasing. IE snyk.io
> 
> This is because lockfiles was made available with 7.0 at 9 April 2021 so for 6 years and 2 months of those 8 years this feature wasn't available. Added to that, IDEA didn't really support it in the beginning, so maybe bump that to the first 7 of your 8 years. I know it's underused, but that doesn't change anything. It is the only way to avoid false positives, just wanted note that, could be a future improvement. Hopefully some future supply chain poisoning with characteristics that would make it preventable with lockfiles and [dependency verification](https://docs.gradle.org/current/userguide/dependency_verification.html) happens, so broad adoption happens.

Was it really only so recently? Surprised but does explain it.
Makes me wonder the format difference between nebula and gradle's built in lockfile is.
https://github.com/nebula-plugins/gradle-dependency-lock-plugin

That one has been available for a long time, though likely hard to say how widely used it is outside of netflix.

deivid-rodriguez commented:
I'll be working on this PR next week. Will push up my initial review suggestions, test out the code here a bit more, and in general try move this PR forward.

alcere commented:
> I'll be working on this PR next week. Will push up my initial review suggestions, test out the code here a bit more, and in general try move this PR forward.

@deivid-rodriguez sorry for taking so long to get you back 😓! I've been quite busy all this time, do you mind if address your comments?I had some spare time this week so I'd like to keep on working on this pull request.

alcere commented on gradle/lib/dependabot/gradle/file_fetcher.rb:
```diff
@@ -127,6 +139,10 @@ def buildfile(dir)
       end
 
+      def gradle_toml_file(dir)
+        find_first(dir, SUPPORTED_VERSION_CATALOG_FILE_PATH) || return
```
Done here: https://github.com/dependabot/dependabot-core/pull/6249/commits/5d1979bdb04614c9475a8163ddfa8685f3dac549

alcere commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -62,6 +67,79 @@ def self.find_includes(buildfile, dependency_files)
+
+          # Only support basic version and reference formats for now,
+          # refrain from updating anything else as it's likely to be a very deliberate choice.
+          next if (version.is_a?(Hash) && (version.length != 1 || version["ref"].nil?)) || version.include?("[")
```
Done here: https://github.com/dependabot/dependabot-core/pull/6249/commits/c9c621ab72dbd7e280d171e0409678088898ee97

alcere commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -62,6 +67,79 @@ def self.find_includes(buildfile, dependency_files)
+
+      def parsed_toml_file(file)
+        @parsed_file ||= {}
+        @parsed_file[file.name] ||= TomlRB.parse(file.content)
```
Makes sense! Done here: https://github.com/dependabot/dependabot-core/pull/6249/commits/d9f39c68b6d49285fb12e36bfa264227d5087c82

deivid-rodriguez commented:
> @deivid-rodriguez sorry for taking so long to get you back 😓! I've been quite busy all this time. Do you mind if I address your comments? I have some spare time this week so I'd like to keep on working on this pull request.

No worries and great having you back in this PR :), thanks for addressing my comments. I see that the PR appears as out of date and a lot of old commits appear in its timeline, I think it needs a rebase?

Can you answer this question from my first review? I lost context now and I haven't yet got to reviewing this again but at the time I was not clear about it:

> My main concern is that when sharing version declarations between multiple dependencies, I think Dependabot may create a different PR for each dependency that uses the reference, all of them with the same contents? Is this correct?

alcere commented:
> My main concern is that when sharing version declarations between multiple dependencies, I think Dependabot may create a different PR for each dependency that uses the reference, all of them with the same contents? Is this correct?

That's a great question @deivid-rodriguez , do you mean something like this?

```
[versions]
composeUi = "1.2.0"

[libraries]
compose-ui = { module = "androidx.compose.ui:ui", version.ref = "composeUi" }
compose-ui-tooling = { module = "androidx.compose.ui:ui-tooling-preview", version.ref = "composeUi" }
test-compose-ui = { module = "androidx.compose.ui:ui-test-junit4", version.ref = "composeUi" }
test-compose-ui-tooling = { module = "androidx.compose.ui:ui-tooling", version.ref = "composeUi" }
test-compose-manifest = { module = "androidx.compose.ui:ui-test-manifest", version.ref = "composeUi" }
```

deivid-rodriguez commented:
Yes, I think that's what I meant.

alcere commented:
> Yes, I think that's what I meant.

You are very right, it does exactly that, I will look into the issue, thanks for the heads up!

deivid-rodriguez commented:
@alcere I had another look at this today and I think this is actually fine. We already have logic in place in the Gradle ecosystem to be able to group PRs together when they are updating a "property" so I think this is ok as is! 🎉

I also reviewed the PR, cleaned up git history and fixed a few bugs while I was at it. Were you planning on pushing any more changes? If not, I can push my cleanup and I think this would be mostly ready! 💪

alcere commented:
> @alcere I had another look at this today and I think this is actually fine. We already have logic in place in the Gradle ecosystem to be able to group PRs together when they are updating a "property" so I think this is ok as is! 🎉
> 
> I also reviewed the PR, cleaned up git history and fixed a few bugs while I was at it. Were you planning on pushing any more changes? If not, I can push my cleanup and I think this would be mostly ready! 💪

That's great news! Please go ahead!

deivid-rodriguez commented:
I pushed my changes. These include:

* Cleanup git history into a handful of meaningful commits (removed style amendment commits and things like that).
* Made some code simplifications and some style changes to make the implementation more "standard Ruby". 
* I added support for dependency libraries specified as `group:artifact:version` or dependency plugins specified as `id:version`.
* I fixed an issue where Dependabot would crash when dependencies included the string "version" or the string "ref".

alcere commented on gradle/lib/dependabot/gradle/file_parser.rb:
```diff
@@ -88,15 +88,13 @@ def dependencies_for_declarations(declarations, toml_file, details_getter)
         declarations.each do |_mod, declaration|
-          version = declaration["version"]
-          next if version.nil?
+          group, name, version = send(details_getter, declaration)
```
TIL bout send, nice Ruby method

yeikel commented:
If possible, let's please add a test case to cover this https://github.com/dependabot/dependabot-core/issues/6742

deivid-rodriguez commented:
@yeikel The error you reported at #6742 was indeed still happening. I added a spec to make sure dependencies specified like that are ignored for now, and don't cause a crash in the update process like the one reported.
